File: dotnet/src/Services/InvestigationService.cs
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Options;
using System.Text.Json;
using System.Text.Json.Serialization;

/// <summary>
/// Orchestrates an incident investigation by:
/// <list type="number">
///   <item>Checking the in-memory cache for a previously computed result.</item>
///   <item>Building the LLM prompt via the static PromptHelper on a cache miss.</item>
///   <item>Sending the prompt to the configured ILlmService.</item>
///   <item>Deserializing the raw JSON response into a typed InvestigationResult.</item>
///   <item>Caching the result for subsequent requests for the same incident.</item>
/// </list>
/// </summary>
public class InvestigationService : IInvestigationService
{
    /// <summary>
    /// The prompt instructs Claude to respond with camelCase JSON keys.
    /// Case-insensitive matching handles any minor casing variation from the LLM.
    /// JsonStringEnumConverter maps the string enum value (e.g. "MissingIngestionAuditEvent")
    /// to the RootCause enum during LLM response deserialization.
    /// </summary>
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        Converters = { new JsonStringEnumConverter() }
    };

    private readonly ILlmService                   _llm;
    private readonly IMemoryCache                  _cache;
    private readonly CacheOptions                  _cacheOptions;
    private readonly ClaudeOptions                 _claudeOptions;
    private readonly ILogger<InvestigationService> _logger;

    public InvestigationService(
        ILlmService llm,
        IMemoryCache cache,
        IOptions<CacheOptions> cacheOptions,
        IOptions<ClaudeOptions> claudeOptions,
        ILogger<InvestigationService> logger)
    {
        _llm           = llm;
        _cache         = cache;
        _cacheOptions  = cacheOptions.Value;
        _claudeOptions = claudeOptions.Value;
        _logger        = logger;
    }

    /// <inheritdoc/>
    public async Task<(InvestigationResult Result, bool FromCache, LlmCompletionResult? LlmMetadata)> InvestigateAsync(
        EventLog eventLog,
        bool bypassCache = false,
        CancellationToken ct = default)
    {
        var cacheKey = $"investigation:event:{eventLog.EventId}";

        if (!bypassCache && _cacheOptions.Enabled &&
            _cache.TryGetValue(cacheKey, out InvestigationResult? cached) && cached is not null)
        {
            _logger.LogInformation("Cache HIT for EventId={EventId}", eventLog.EventId);
            return (cached, FromCache: true, LlmMetadata: null);
        }

        // 1. Build prompt — pure static helper, no injection needed
        var prompt = PromptHelper.BuildInvestigationPrompt(eventLog, _claudeOptions.MaxPromptChars);

        _logger.LogInformation(
            "Cache MISS — starting LLM investigation for EventId={EventId} ScenarioName={ScenarioName}",
            eventLog.EventId, eventLog.ScenarioMetadata?.ScenarioName ?? "N/A");

        // 2. Call the LLM — provider determined by which ILlmService is registered
        var llmCompletion = await _llm.CompleteAsync(prompt, ct);

        _logger.LogDebug("LLM raw response for EventId={EventId}: {LlmResponse}", eventLog.EventId, llmCompletion.Text);

        // 3. Strip markdown code fences if the LLM wrapped the JSON in ```json ... ```
        var json = llmCompletion.Text.Trim();
        if (json.StartsWith("```", StringComparison.Ordinal))
        {
            var firstNewline = json.IndexOf('\n');
            if (firstNewline >= 0)
                json = json[(firstNewline + 1)..];

            if (json.EndsWith("```", StringComparison.Ordinal))
                json = json[..^3].TrimEnd();
        }

        // 4. Deserialize the structured JSON verdict
        var result = JsonSerializer.Deserialize<InvestigationResult>(json, JsonOptions);

        if (result is null)
            throw new InvalidOperationException(
                $"LLM response could not be deserialized for EventId={eventLog.EventId}. Raw: {llmCompletion.Text}");

        _logger.LogInformation(
            "Investigation complete for EventId={EventId}: RootCause={RootCause} Confidence={Confidence} Model={Model} InputTokens={InputTokens} OutputTokens={OutputTokens}",
            eventLog.EventId, result.RootCause, result.Confidence,
            llmCompletion.Model, llmCompletion.InputTokens, llmCompletion.OutputTokens);

        // 5. Cache the result for subsequent requests (only when caching is enabled and not bypassed)
        if (!bypassCache && _cacheOptions.Enabled)
        {
            var entryOptions = new MemoryCacheEntryOptions
            {
                AbsoluteExpirationRelativeToNow = TimeSpan.FromHours(_cacheOptions.TtlHours),
                Priority = CacheItemPriority.Normal
            };
            _cache.Set(cacheKey, result, entryOptions);
        }

        return (result, FromCache: false, LlmMetadata: llmCompletion);
    }
}

File: python/app/services/investigation_service.py
"""
Orchestrates an incident investigation: builds the LLM prompt, calls the LLM,
deserializes the response, and caches results.
≈ InvestigationService.cs
"""
from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass
from typing import Optional

import jsonschema
from cachetools import TTLCache

from app.helpers.prompt_helper import build_investigation_prompt
from app.models.investigation import InvestigationResult
from app.models.root_cause import RootCause
from app.services.llm_service import ClaudeService, LlmCompletionResult

logger = logging.getLogger(__name__)

# Load LLM response schema for validation
from pathlib import Path
# parents[0] = app/services, parents[1] = app, parents[2] = python/, parents[3] = repo root.
# The shared/ directory lives at the repo root.
_SHARED_DIR = Path(__file__).parents[3] / "shared"
_LLM_SCHEMA: dict = json.loads((_SHARED_DIR / "config" / "llm-response-schema.json").read_text(encoding="utf-8"))


@dataclass
class InvestigationCacheEntry:
    result: InvestigationResult


class InvestigationService:
    """
    Orchestrates incident investigations with optional TTL caching.
    Accepts an EventLog (mirrors .NET InvestigationService which takes EventLog).
    Cache key: "investigation:event:{event_id}" — mirrors .NET.
    ≈ InvestigationService.cs
    """

    def __init__(
        self,
        llm: ClaudeService,
        cache_enabled: bool = False,
        ttl_hours: int = 24,
        max_prompt_chars: int = 600_000,
    ) -> None:
        self._llm              = llm
        self._cache_enabled    = cache_enabled
        self._max_prompt_chars = max_prompt_chars
        self._cache: TTLCache[str, InvestigationResult] = TTLCache(
            maxsize=1000,
            ttl=ttl_hours * 3600,
        )
        self._lock = asyncio.Lock()

    async def investigate_async(
        self,
        event_log: "EventLog",
        bypass_cache: bool = False,
    ) -> tuple[InvestigationResult, bool, Optional[LlmCompletionResult]]:
        """
        Investigates the event log and returns (result, from_cache, llm_metadata).
        When bypass_cache=True the in-memory cache is skipped for both reads and writes.

        Cache key mirrors the .NET implementation:
          "investigation:event:{event_id}"  (InvestigationService.cs:56)
        ≈ InvestigationService.InvestigateAsync()
        """
        from app.models.event_log import EventLog

        cache_key = f"investigation:event:{event_log.event_id}"

        # Check cache (only when caching is enabled and not bypassed)
        if not bypass_cache and self._cache_enabled:
            cached = self._cache.get(cache_key)
            if cached is not None:
                logger.info(
                    "Cache HIT for event_id=%s",
                    event_log.event_id,
                )
                return cached, True, None

        # Build prompt
        prompt = build_investigation_prompt(event_log, self._max_prompt_chars)

        logger.info(
            "Cache MISS — starting LLM investigation for event_id=%s scenario=%s",
            event_log.event_id,
            event_log.scenario_metadata.scenario_name if event_log.scenario_metadata else "N/A",
        )

        # Call LLM
        llm_completion = await self._llm.complete_async(prompt)

        logger.debug("LLM raw response for event_id=%s: %s", event_log.event_id, llm_completion.text)

        # Strip markdown code fences if the LLM wrapped the JSON in ```json ... ```
        # Mirrors InvestigationService.cs lines 78-87 exactly:
        #   1. Trim whitespace.
        #   2. If the string starts with ```, slice off the opening fence line.
        #   3. If the string then ends with ```, slice off the trailing fence.
        raw_json = llm_completion.text.strip()
        if raw_json.startswith("```"):
            first_newline = raw_json.find("\n")
            if first_newline >= 0:
                raw_json = raw_json[first_newline + 1:]
            if raw_json.endswith("```"):
                raw_json = raw_json[:-3].rstrip()

        # Parse and validate against schema
        parsed = json.loads(raw_json)
        jsonschema.validate(parsed, _LLM_SCHEMA)

        result = InvestigationResult(
            root_cause=RootCause(parsed["rootCause"]),
            explanation=parsed["explanation"],
            confidence=parsed["confidence"],
        )

        logger.info(
            "Investigation complete for event_id=%s: root_cause=%s confidence=%s model=%s",
            event_log.event_id,
            result.root_cause,
            result.confidence,
            llm_completion.model,
        )

        # Cache the result (only when caching is enabled and not bypassed)
        if not bypass_cache and self._cache_enabled:
            async with self._lock:
                self._cache[cache_key] = result

        return result, False, llm_completion
